File: src/10.graph/18.491.go
package graph

func FindSubsequences(nums []int) [][]int {
	ans := [][]int{}
	onPath := []int{}
	n := len(nums)

	var backtrack func(r int)

	backtrack = func(r int) {
		if len(onPath) >= 2 {
			ans = append(ans, append([]int{}, onPath...))
		}

		// 去重, 使用过一次之后就不能继续使用了
		used := make(map[int]bool, n)

		for i := r; i < n; i++ {
			if used[nums[i]] {
				continue
			}

			if len(onPath) == 0 || nums[i] >= onPath[len(onPath)-1] {
				onPath = append(onPath, nums[i])
				used[nums[i]] = true
				backtrack(i + 1)
				onPath = onPath[:len(onPath)-1]
			}
		}
	}

	backtrack(0)

	return ans
}

File: src/10.graph/18.491.ts
export function findSubsequences(nums: number[]): number[][] {
    const ans: number[][] = [];
    const onPath: number[] = [];
    const n = nums.length;
    const backtrack = (r: number) => {
        if (onPath.length >= 2) {
            ans.push([...onPath]);
        }

        const used = new Map<number, boolean>();

        for (let i = r; i < n; i++) {
            if (used.has(nums[i])) {
                continue;
            }

            if (onPath.length === 0 || nums[i] >= onPath[onPath.length - 1]) {
                used.set(nums[i], true);
                onPath.push(nums[i]);
                backtrack(i + 1);
                onPath.pop();
            }
        }
    };

    backtrack(0)

    return ans;
}
